src/contracts/middleware/Middleware.sol

```diff
@@ -219,18 +263,54 @@ contract Middleware is
     // /**
     //  * @inheritdoc IMiddleware
     //  */
-    //  TODO: this function should be split to allow to be called by chainlink in case
-    function sendCurrentOperatorsKeys() external returns (bytes32[] memory sortedKeys) {
-        if (address(s_gateway) == address(0)) {
-            revert Middleware__GatewayNotSet();
-        }
-
+    function sendCurrentOperatorsKeys() external onlyIfGatewayExists returns (bytes32[] memory sortedKeys) {
         uint48 epoch = getCurrentEpoch();
         sortedKeys = sortOperatorsByVaults(epoch);
 
         s_gateway.sendOperatorsData(sortedKeys, epoch);
     }
 
+    /**
+     * @inheritdoc AutomationCompatibleInterface
+     * @dev Called by chainlink nodes off-chain to check if the upkeep is needed
+     * @return upkeepNeeded boolean to indicate whether the keeper should call performUpkeep or not.
+     * @return performData bytes of the sorted operators' keys and the epoch that will be used by the keeper when calling performUpkeep, if upkeep is needed.
+     */
+    function checkUpkeep(
+        bytes calldata /* checkData */
+    ) external view override returns (bool upkeepNeeded, bytes memory performData) {
+        uint48 epoch = getCurrentEpoch();
+        bytes32[] memory sortedKeys = sortOperatorsByVaults(epoch);
+
+        // Should it be epochDuration? Because we wanna send it once per network epoch
+        upkeepNeeded = (block.timestamp - s_lastTimestamp) > s_interval;
+
+        performData = abi.encode(sortedKeys, epoch);
+
+        return (upkeepNeeded, performData);
+    }
+
+    /**
+     * @inheritdoc AutomationCompatibleInterface
+     * @dev Called by chainlink nodes off-chain to perform the upkeep. It will send the sorted keys to the gateway
+     */
+    function performUpkeep(
+        bytes calldata performData
+    ) external override onlyIfGatewayExists {
+        if (msg.sender != s_forwarderAddress) {
+            revert Middleware__NotForwarder();
+        }
+
+        if ((block.timestamp - s_lastTimestamp) > s_interval) {
+            s_lastTimestamp = block.timestamp;
+
+            // Decode the sorted keys and the epoch from performData
+            bytes32[] memory sortedKeys = abi.decode(performData, (bytes32[]));
```

Comment:
On the `checkUpKeep` you encode `keys + epoch`, but here you are only decoding the first. Is that ok?